keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+component should be annotated with a reference to the Application.
+
+An Application's `Spec.Dependencies` is a list of other Applications on which the application depends. Clients can use 
+this field to discover and retrieve other Applications that the containing Application requires to be present in the 
+system.
+
+While we attempt to represent many properties of the Application object with explicit fields, we realize that the 
+current list is not exhaustive, and we do not endeavor to develop an exhaustive list prior to releasing the 
+Application Kind. We expect users to use the fields of ObjectMeta (i.e. Labels and Annotations) for custom use cases. 
+When the community forms a consensus that a particular label or annotation has become prevalent across a large 
+number of tools, we will promote that metadata to a field. Note that this can be done in a backward compatible manner 
+without respect to the stability of the Application Kind.
+
+#### Validation
+Upon Application creation, The API Server shall perform the following validation.
+
+1. The API Server shall validate that the Application is a valid ObjectMeta with a name that is a valid DNS subdomain. 
+In effect, the API Server will simply validate that the Application is a valid Kubernetes resource.
+1. The API Server will validate that the Application's `Spec.Selector` is a valid, non-nil, set-based Selector.
+1. If the Application has a `.Spec.HealthCheck`, the API Server will validate that the `Template` of the 
+ApplicationHealthCheck is a valid Pod with a readiness check.
+
+The API Server shall perform the following validation upon Application mutation.
+
+1. The API server shall ensure that the new version of the Application is valid using the above definition of a valid 
+newly created Application instance. 
+1. The API server shall ensure that the Applications `Spec.Selector` has not been mutated. This latter condition has 
+been implemented in many places in the Kubernetes API to prevent unintentional orphaning. This is our purpose here.
+
+#### Generation Incrementation
+
+1. Upon creation of an Application object, the API Server shall set the initial value of the `Generation` field to 0.
+1. The API Server shall, upon update of an Application object, increment the `Generation` of the Application object by 1. 
+
+
+#### Namespaces
+All components and dependencies of an Application must reside in the same namespace as the Application. Given the 
+design of Kubernetes namespaces, we feel this is a reasonable, and desirable, constraint. A user, who has write 
+permissions for a particular namespace, can install an Application and all of its components into the same namespace, 
+and only those users with the necessary permissions can modify the Application. Additionally, administrators may limit 
+the resources consumed by an Application by applying quotas to the namespace in which it resides.
```

Comment:
I understand the boundary for members of the app, but I don't understand why this boundary for dependencies. I'm getting hazier and hazier on what a dependency is supposed to model as I read through this document. If a dependency is strongly coupled, then imposing namespace restrictions makes sense. But a weakly coupled dependency -- and particularly a shared dependency -- probably shouldn't be subjected to the same constraints.